controller_interface/src/chainable_controller_interface.cpp

```diff
@@ -44,6 +44,42 @@ return_type ChainableControllerInterface::update(
   return ret;
 }
 
+std::vector<hardware_interface::StateInterface>
+ChainableControllerInterface::export_state_interfaces()
+{
+  auto state_interfaces = on_export_state_interfaces();
+  // check if the "state_interfaces_values_" variable is resized to number of interfaces
+  if (state_interfaces_values_.size() != state_interfaces.size())
+  {
+    RCLCPP_FATAL(
+      get_node()->get_logger(),
+      "The internal storage for exported state values 'state_interfaces_values_' variable has size "
+      "'%zu', but it is expected to have the size '%zu' equal to the number of exported state "
+      "interfaces. No state interface will be exported. Please correct and recompile "
+      "the controller with name '%s' and try again.",
+      state_interfaces_values_.size(), state_interfaces.size(), get_node()->get_name());
+    state_interfaces.clear();
+  }
+
+  // check if the names of the controller state interfaces begin with the controller's name
+  for (const auto & interface : state_interfaces)
+  {
+    if (interface.get_prefix_name() != get_node()->get_name())
+    {
+      RCLCPP_FATAL(
+        get_node()->get_logger(),
+        "The name of the interface '%s' does not begin with the controller's name. This is "
+        "mandatory for state interfaces. No state interface will be exported. Please "
+        "correct and recompile the controller with name '%s' and try again.",
+        interface.get_name().c_str(), get_node()->get_name());
+      state_interfaces.clear();
+      break;
+    }
+  }
+
+  return state_interfaces;
```

Comment:
The code to generate the *almost* one and only implementation that meets this criteria in a meaningful way takes a vector of strings, pairs them with the values in `state_interfaces_values_` and creates a vector of `StateInterface`s. I think we should not make people copy that piece of code around but instead generate it right here, with the right prefix and everything.

Reply:
In my mind both values and names should be set up in a single place, either the constructor or `init` or `configure`, the state interface export can be implemented by us.

Reply:
I agree with you about having methods that do it automatically for the user. I would make it optional as in the user can decide to use those helper methods or not. I personally don't like to force users to use this specific data type. For instance, one of my state interfaces is the position of the base of the robot, and sometimes it is better to have it `Eigen::Vector3d` or `geometry_msgs/Point`. This would give some flexibility to the user in my opinion.

If it is only the above position it is not a problem, but if he has orientation and all joints state etc, then it's a huge chunk of data and it gets messy when the user wants to remap the data into his own struct.

What do you think about it?

Reply:
What if we left this check but had the suggested default implementation too? Still erring on the side of not making people copy&paste the same boilerplate code all the time.

Reply:
Sure, we can do that. I can do the boilerplate part, and we can still do error on this side. I like it

Reply:
I've just implemented it by adding a member variable and the user fills in the list of variable names to be exported in the configure stage etc and it will done automatically. 
Take a look at the implementation and let me know.

Thank you :)

Reply:
Thank you!